### ISSUE TITLE
Implement validators for core obspy classes

### ISSUE THREAD

**Issue by d-chambers** [obspy/obspy#2154]
I am opening this issue to discuss implementing validation methods on the `Stream`, `Catalog`, and `Inventory` classes. Possible use of validators have been discussed in #1117 and #2091 (maybe others).

There are several reasons why a validation method might be useful:

1. Ensure format specific requirements are met before attempting to write data to a particular format. 

2. Allow users to implement custom validation functions that can be called before specific use cases.

_API_

Registering validators could be done in a declarative manner with decorators. There would probably be three decorators: `validate_catalog`, `validate_inventory` and  `validate_stream`. Each validator would take two optional arguments: `name` and `cls`. The parameter `name` will simply provide a means for namespacing validators and selectively controlling their execution. 

The parameter 'cls' will specify a class (`Catalog`, `Inventory`, `Stream`, `Response`, `Event`, `Magnitude`, `Trace`, etc.) which the decorated function expects as the only input. Doing so will allow the validation implementation to recursively parse the object tree as single time and apply validators to the appropriate class instances when they are found. An example is probably in order:  

_Example_

Suppose we are to use python bindings from the program hypoinverse to locate a catalog of events, I might implement the following validators to ensure the `Catalog` instance, and other class instances attached as potentially nested attributes, are valid input for the location function:  

```python
import obspy
from obspy.core.event import Event, Pick

# define validation functions for the 'hypoinverse' namespace
@validate_catalog(name='hypoinverse', cls=Event)
def ensure_four_picks(event):
    assert len(event.picks) >= 4


@validate_catalog(name='hypoinverse', cls=Pick)
def ensure_picks_have_time(pick):
    assert isinstance(pick.time, obspy.UTCDateTime)


@validate_catalog(name='hypoinverse', cls=Pick)
def ensure_picks_have_phase_hint(pick):
    assert pick.phase_hint is not None


@validate_catalog(name='hypoinverse')  # because no cls is provided the default is used
def ensure_at_least_one_event(catalog):
    assert len(catalog)


cat = obspy.read_events('some_catalog.xml')
cat.validate('hypoinverse')  # raises an AssertionError if any of the expectations are not met
```
_Implementation_

A global registry of validators would be kept for each core data structure. The registries might look like this:

```pythpon
_CATALOG_VALIDATORS = {'namespace1': {cls1: [validator1, validator2, ...]}
_STREAM_VALIDATORS = ...
_INVENTORY_VALIDATORS = ...
```
The `validate_X` decorators would then simply registry the decorator function in the appropriate place. 

The validation methods would perform the following when called:

1. Ensure all validation plug-ins of the desired type have been loaded. This will allow validators to be registered with plugins just as IO methods are.

2. Extract the validators associated with the input name by indexing the validator registry with supplied `name` parameter.

3. Recurse the object tree and run validators when instance of classes with registered validators are found. 

_Pros_
* simple, declarative, API
* supports registering suites of validators for various purposes (not just as pre-IO checks)
* avoids excessive iteration of core objects to implement multiple validators
* shields user from the complexity of object transversal

_Cons_
* Many users new to python may not be familiar with decorators
* Validators would only be applied to instances of registered classes, not their subclasses. Making it work with subclasses would be possible but more complicated. 

Thoughts?

**Comment by calum-chamberlain** [comment]
I like this for the pros, particularly shielding people from the inner workings of objects. However, is calling external things (such as hypoinverse) something that you see coming under obspy at some point (I would love to have some kind of location algorithm within obspy, but it feels somewhat of a stretch from where obspy currently resides). 
If not, then maybe this kind of thing is better applied for those applications that require it? 
I don't have strong opinions either way, but thought I would chime in. To me it seems like a good idea, especially if obspy plans on growing in this way. Are you already making use of these validators in other ways @d-chambers?

**Comment by d-chambers** [comment]
Hey @calum-chamberlain ,

The hypoinverse example may have been a bit distracting. I think the primary purpose of this, at least internal to obspy, would be for data validation before trying to write to a particular format. IMHO it would be better to release a small package for using hypoinverse, or another location code, rather than trying to stuff it into obspy. 

I do think having a built-in way to register and run validators would be useful for external packages, mainly so that each package doesn't have to implement its own way of doing it. You maintain such a package, do you think something like this would be useful for EQCorrScan? Perhaps you would register a few expectations for input catalogs (maybe that they have some picks?) then you could call the `validate` method to make sure inputs are valid before getting into any business logic. This way EQCorrScan could simply declare what it expects of input data and leave the object traversal to the obspy machinery. 

I have already implemented something similar at work and it has worked out well so far.

**Comment by calum-chamberlain** [comment]
It would certainly be useful for EQcorrscan - but at the moment things are handled by errors within EQcorrscan (although those error messages could be better). Are you proposing that obspy would just allow the ability to declare validators, and that other people would write them for their use-case (which seems sensible to me?)?
Sounds good to me.

Apologies for getting distracted by the hypoinverse thing - I quite want to write some location routines in Python at some point, it would be sweet to make an `Event.locate(method="...", ...)` type thing, but that should definitely be done outside obspy.

**Comment by megies** [comment]
> it would be sweet to make an Event.locate(method="...", ...) type thing, but that should definitely be done outside obspy.

On the danger of going off-topic.. this indeed would be very nice, but it needs some brain efforts for a good design. Maybe we could implement an abstract interface of a `Locator` object with the expected routines inside obspy and then enable adding them via plugins?

**Comment by d-chambers** [comment]
>>> Are you proposing that obspy would just allow the ability to declare validators, and that other people would write them for their use-case

Exactly. Obspy would use the validator logic internally and why not expose it to make it easier for obspy dependent packages to use as well? This should be fine as long as it stays organized through namespacing. 

Taking us off topic a bit more... 

>>> Maybe we could implement an abstract interface of a Locator object ...

That is not a bad idea, but I think we already have the data structures to define a good interface for a location code:

```python
from obspy import Inventory
from obspy.core.event import Origin, Event


def locate_event(event: Event, inventory: Inventory) -> Origin:
    pass
```

It may need some other arguments as well, such as a config dictionary and maybe a velocity model. It is not yet clear to me how a velocity model should be defined in python but I think [xarray](https://xarray.pydata.org/en/stable/) holds promise. 

Anyway, no sense in really spending much time discussing location interfaces if there isn't a need (yet).

**Comment by calum-chamberlain** [comment]
> Exactly. Obspy would use the validator logic internally and why not expose it to make it easier for obspy dependent packages to use as well? This should be fine as long as it stays organized through namespacing.

:+1: sounds grand to me, I would use it.

---

Continuing off topic for the hell of it:

I was thinking something more like this that subclasses `Event` (just because I like the idea of being able to use all of the methods of `Event` immediately):

```python
from obspy import Inventory
from obspy.core.event import Event

class Locator(Event):
   def locate(method: str, inventory: Inventory, v_model: xarray):
      origin = do_location()
      self.origins.extend(origin)
```
Agree that obspy has a lot of the structures required for this, but I'm still not convinced obspy is the place for it - being able to plug Locators in could be really nice.

Anyway, I shall say no more other than I would really like to see some integrated way to locate events from within Python with some nice documentation... I doubt I will find time to do it though :disappointed:

**Comment by megies** [comment]
Alright, to separate the thoughts on event/location stuff, I created a new issue #2168.

**Comment by dsentinel** [comment]
@d-chambers thanks for starting this thread.

I think this is a great idea!

I'll try to take some time to play with implementation ideas as well.

I tend to avoid decorators unless there's absolutely not a better way, they tend to make code harder to follow, maintain and debug. This may indeed  be a good case for them. It's definitely a nice way to register functions.

`assert...` vs. `return False`:
We don't want failed validation to bubble to the top and stop program execution.  We will also want all validations to run past failed ones. Similar in many ways to how unittests run, but unittests have nice fancy framework that allows this to happen with asserts. Perhaps we can borrow code from a test framework and use asserts similarly.

namspacing is definetly great here. I imagine we'll want a standard format validator, but then users may have a more stringent set of rules say for a particular archive.

**Comment by dsentinel** [comment]
> @dsentinel, thanks for the input. A lot of this was inspired by pytest
> which registers fixtures through decorator and which uses plain
> asserts in tests statements. 

Right, this would be really slick.  Unit tests are strikingly similar to this problem. 
There's a lot under the hood to make test runners so slick; re-writing assert on imports, inspecting the stack and frame, etc. Maybe overkill, or a lot of work, but awesome if we got it.
> 
> > We don't want failed validation to bubble to the top and stop program execution
> 
> Sometimes we might; if there is a serious problem it may be better to
> fail fast. I agree, however, this may not always be the case. The
> logic that runs the validators could catch  the `AssertionError` and
> decide if it should be converted into a warning to re-raised. Maybe
> something like this?

I agree that a fast fail option is useful, and, after more thought, that raising exceptions have cases.

```python
import obspy
cat = obspy.read_events()
cat.validate('some_namespace', raise=True)  # re-raise assertion errors
cat.validate('some_namespace', raise=False)  # convert assertion errors to warnings
```

> Likewise, returning `True` or `False` could work but it doesn't allow
> the user to attach a message and doesn't quite look as, well,
> assertive?

I'd like to see more than just a message. I think it'd be good to get back the offending object at least
Perhaps:
```python
valid = cat.validate(critical_namespace, fast_fail=True):
if not valid:
    print(valid.failures)		# Shows only one error
    sys.exit(1)
valid = cat.validate(warning_namespace, fast_fail=False)
if not vaild:
    print(valid.failures)	# Shwo all the errors
    for error_msg, offending_objects, bad_value, namespace in valid.falures:
        ...
	# Do stuff with marginal cat
# do stuff with happy cat, like pet it
```
where `validate()` returns True or a failure object similar to how the re module `search()` returns a match object or None, but with truthiness flipped.... It may be better to flip the verbs to avoid flipping where
```python
failures = cat.invalidate()
if failures:
	...
```
> 
> > We will also want all validations to run past failed ones.
> 
> This would be a potentially useful feature but may add a lot of
> complexity. Storing the state of past runs for each (mutable) object
> in a sensible way wouldn't be trivial, especially if the object
> changes between validation runs. I envision simply running all the
> validators registered to a particular namespace every time a
> `validate` method is called to keep things as simple as possible.

I think we're on the same page here. I'm not implying state, I was referring to a single call to `validate()` returning all the failures which is hard to do if exceptions are coming up. 

### Other things to think about:
- Overloading `__setattr__()` on objects to validate when setting values
- Reporting line numbers of a file when just reading and validating a file. Maybe not implemented here, but structured so that it can be done
- Validation on fields across objects/tree i.e. `Inventory.network.start <= channel.start`

### My perspective/use case/angle
I developed a GUI stationXML creator/editor. I would like to:
 - Validate an inventory prior to save. Then highlight the failures in the GUI
 - Warn reject invalid user input on form fields at input time.
 - 
I also have a pre archive qc system that would benifit from this, but I think the use case would be similar to the typical.

**Comment by d-chambers** [comment]
Hey @dsentinel,

Good idea. I like the idea of the `validate` method returning a `Failures` object which contains a list of `Failure` instances in the `failures` attribute. The `__bool__` would return True if any failures were found and False otherwise. The 'Failue` class would of course have a nice `__str__` output to convey line numbers, assert messages, validation function name, etc.

Here is the recursive function for yielding non built-in objects from object trees: https://github.com/obspy/obspy/blob/53bf7e701df3d872aa62c4e4fff7dbcf759942e5/obspy/core/util/misc.py#L641-L708. It would give us the basic machinery to test all instances of specified classes, which I think is what you mean when you say "validation on fields across objects/tree".  It could also give us the ability to test all attributes with a certain name (eg any attr on any object in the tree named "latitude" should be gt -90 and lt 90).

> Overloading __setattr__() on objects to validate when setting values
I think in the spirit of what we were discussing in #1117 we should avoid doing this as much as possible. Better to simply check if an object is a valid input to some operation before performing said operation. In some cases it may make sense but that would probably be object specific and outside of the scope of this PR.  

I may not have much time to work on an implementation for a few weeks but if you want to get started feel free.

**Comment by dsentinel** [comment]
> Here is the recursive function for yielding non built-in objects from object trees:
...

Nice idea. I think this would it.

> Overloading setattr() on objects to validate when setting values
> 
> I think in the spirit of what we were discussing in #1117 we should avoid doing this as much as possible. Better to simply check if an object is a valid input to some operation before performing said operation. In some cases it may make sense but that would probably be object specific and outside of the scope of this PR.

I don't think overloading `settattr()` should be done in this PR. I'd like to see this PR have an API / path that would allow for such.

> I may not have much time to work on an implementation for a few weeks but if you want to get started feel free.

No worries. I'm still have some things to do before I can crank on this as well.

**Comment by dsentinel** [comment]
I'm not sure it's the best way to go, but there is overlap, and this concept isn't new. It might be worth looking into a traits, or traitlets for ideas if not use.
